File: code-sample-java/code-sample-spring-security/src/main/java/learn_to_code/springboot/security/DbUserDetailsService.java
package learn_to_code.springboot.security;

import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;


@Service
public class DbUserDetailsService implements UserDetailsService {

    private UserRepository userRepository;

    public DbUserDetailsService(UserRepository repository) {
        userRepository = repository;
    }

    @Override
    public UserDetails loadUserByUsername(String userName) throws UsernameNotFoundException {
        return userRepository.findByUsername(userName).orElseThrow(UserNotFoundException::new);
    }
}

File: code-sample-kotlin/spring-boot/code-sample-kotlin-spring-security/src/main/kotlin/codesample/kotlin/service/DbUserDetailsService.kt
package codesample.kotlin.service

import codesample.kotlin.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service

@Service
class DbUserDetailsService (@Autowired private val userRepository: UserRepository) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails
        = userRepository.findByUsername(username) ?: throw UsernameNotFoundException("user $username not found")
}